lib/dynasm-backend/src/lib.rs

```diff
@@ -0,0 +1,302 @@
+use std::ptr::NonNull;
+use std::sync::Arc;
+use wasmer_runtime_core::{
+    backend::{Backend, Compiler, FuncResolver, ProtectedCaller, Token},
+    error::{CompileError, CompileResult, RuntimeResult},
+    module::{
+        DataInitializer, ExportIndex, ImportName, ModuleInfo, ModuleInner, StringTable,
+        TableInitializer,
+    },
+    structures::{Map, TypedIndex},
+    types::{
+        ElementType, FuncIndex, FuncSig, GlobalDescriptor, GlobalIndex, GlobalInit,
+        ImportedGlobalIndex, Initializer, LocalFuncIndex, MemoryDescriptor, MemoryIndex,
+        TableDescriptor, TableIndex, Type as CoreType, Value,
+    },
+    units::Pages,
+    vm::{self, ImportBacking},
+};
+use wasmparser::{
+    self, ExternalKind, FuncType, ImportSectionEntryType, InitExpr, MemoryType, ModuleReader,
+    Operator, SectionCode, TableType, Type, WasmDecoder,
+};
+
+struct Placeholder;
+
+impl FuncResolver for Placeholder {
+    fn get(
+        &self,
+        _module: &ModuleInner,
+        _local_func_index: LocalFuncIndex,
+    ) -> Option<NonNull<vm::Func>> {
+        None
+    }
+}
+
+impl ProtectedCaller for Placeholder {
+    fn call(
+        &self,
+        _module: &ModuleInner,
+        _func_index: FuncIndex,
+        _params: &[Value],
+        _import_backing: &ImportBacking,
+        _vmctx: *mut vm::Ctx,
+        _: Token,
+    ) -> RuntimeResult<Vec<Value>> {
+        Ok(vec![])
+    }
+}
+
+pub struct DynasmCompiler {}
+
+impl DynasmCompiler {
+    pub fn new() -> DynasmCompiler {
+        DynasmCompiler {}
+    }
+}
+
+impl Compiler for DynasmCompiler {
+    fn compile(&self, wasm: &[u8], _: Token) -> CompileResult<ModuleInner> {
+        validate(wasm)?;
```

Comment:
Instead of validating and then parse the module, can you do both at the same time? We want to be able to support streaming compilation in the future.